File: 2020/04/run04.go
package main

import (
  "io/ioutil"
  "fmt"
  "os"
  "strings"
)

type passport struct {
  byr string
  iyr string
  eyr string
  hgt string
  hcl string
  ecl string
  pid string
  cid string
}

var example_passwords_input = `ecl:gry pid:860033327 eyr:2020 hcl:#fffffd
byr:1937 iyr:2017 cid:147 hgt:183cm

iyr:2013 ecl:amb cid:350 eyr:2023 pid:028048884
hcl:#cfa07d byr:1929

hcl:#ae17e1 iyr:2013
eyr:2024
ecl:brn pid:760753108 byr:1931
hgt:179cm

hcl:#cfa07d eyr:2025 pid:166559648
iyr:2011 ecl:brn hgt:59in
`

func extractPassportLines(passports_blob string) []string {
  var (
    index int
    passport_lines []string
    passport_line string
  )
  passport_lines = strings.Split(passports_blob, "\n\n")
  for index, passport_line = range passport_lines {
    passport_lines[index] = strings.Trim(passport_line, "\n")
  }
  return passport_lines
}

func extractPassportData(password_line string) (*passport, error) {
  var (
    pass passport
    password_items []string
    key_value []string
    item string
    key string
    value string
  )
  password_line = strings.Trim(password_line, "\n")
  password_line = strings.ReplaceAll(password_line, "\n", " ")
  password_items = strings.Split(password_line, " ")
  for _, item = range password_items {
    key_value = strings.SplitN(item, ":", 2)
    if len(key_value) != 2 {
      return &pass, fmt.Errorf("Fail to extract key:value pair from '%s' in %s", item, password_line)
    }
    value = key_value[1]
    switch key = key_value[0]; key {
    case "byr":
      pass.byr = value
    case "iyr":
      pass.iyr = value
    case "eyr":
      pass.eyr = value
    case "hgt":
      pass.hgt = value
    case "hcl":
      pass.hcl = value
    case "ecl":
      pass.ecl = value
    case "pid":
      pass.pid = value
    case "cid":
      pass.cid = value
    default:
      return &pass, fmt.Errorf("Unkown key '%s' with value '%s' from '%s'", key, value, password_line)
    }
  }
  return &pass, nil
}

func checkPassportRequiredField(pass *passport) bool {
  if pass.byr != "" && pass.iyr != "" && pass.eyr != "" && pass.hgt != "" && pass.hcl != "" && pass.ecl != "" && pass.pid != "" {
    return true
  }
  return false
}

func checkByr(byr_value string) bool {
  return true
}

func checkHgt(hgt_value string) bool {
  return true
}

func checkHcl(hcl_value string) bool {
  return true
}

func checkEcl(ecl_value string) bool {
  return true
}

func checkPid(pid_value string) bool {
  return true
}


func commonOne(passports_blob string) int {
  var (
    valid_passports_number int
    passport_lines []string
    passport_line string
    pass *passport
    err error
  )
  passport_lines = extractPassportLines(passports_blob)
  for _, passport_line = range passport_lines {
    pass, err = extractPassportData(passport_line)
    if err == nil {
      if checkPassportRequiredField(pass) {
        valid_passports_number += 1
      }
    } else {
      fmt.Printf("Error: %s\n", err)
    }
  }
  return valid_passports_number
}

func exampleOne() {
  var valid_passports_number int
  valid_passports_number = commonOne(example_passwords_input)
  fmt.Printf("ExampleOne: %v\n", valid_passports_number)
}


func mainOne() {
  var (
    valid_passports_number int
    input []byte
    err error
  )
  input, err = ioutil.ReadFile("input4.txt")
  if err == nil {
    valid_passports_number = commonOne(string(input))
    fmt.Printf("MainOne: %v\n", valid_passports_number)
  } else {
    fmt.Printf("MainOne: I/O erro : %s\n", err)
  }
}

func commonTwo(passports_blob string) int {
  var (
    valid_passports_number int
    passport_lines []string
    passport_line string
    pass *passport
    err error
  )
  passport_lines = extractPassportLines(passports_blob)
  for _, passport_line = range passport_lines {
    pass, err = extractPassportData(passport_line)
    if err == nil {
      if checkPassportRequiredField(pass) {
        valid_passports_number += 1
      }
    } else {
      fmt.Printf("Error: %s\n", err)
    }
  }
  return valid_passports_number
}

func exampleTwo() {
  var valid_passports_number int
  valid_passports_number = commonOne(example_passwords_input)
  fmt.Printf("ExampleTwo: %v\n", valid_passports_number)
}


func mainTwo() {
  var (
    valid_passports_number int
    input []byte
    err error
  )
  input, err = ioutil.ReadFile("input4.txt")
  if err == nil {
    valid_passports_number = commonOne(string(input))
    fmt.Printf("MainTwo: %v\n", valid_passports_number)
  } else {
    fmt.Printf("MainTwo: I/O erro : %s\n", err)
  }
}

func main() {
  request := "all"
  if len(os.Args) > 1 {
    request = os.Args[1]
  }
  if request == "1" || request == "example1" || request == "all" {
    exampleOne()
  }
  if request == "1" || request == "input1" || request == "all" {
    mainOne()
  }
  if request == "2" || request == "example2" || request == "all" {
    exampleTwo()
  }
  if request == "2" || request == "input2" || request == "all" {
    mainTwo()
  }
}

File: 2020/04/run04.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import functools
import itertools
import operator
import pathlib
import re

from pprint import pprint as print

COLOR_RE = re.compile(r'#[0-9a-f]{6}')

def extract_password_data(password_line: str) -> 'dict[str,str]':
    password_items = password_line.split(' ')
    password_items = (item.split(':', 1) for item in password_items)
    password_items = {k: v for k, v in password_items}
    return password_items

def read_passwords():
    input_file = pathlib.Path('input4.txt')
    input_file_data = input_file.read_text()
    password_blocks = input_file_data.split('\n\n')
    password_blocks = (line.strip('\n') for line in password_blocks)
    password_lines = (line.replace('\n', ' ') for line in password_blocks)
    password_items = map(extract_password_data, password_lines)
    return password_items

REQUIRED_FIELDS = {
    'byr': lambda v: 1920 <= int(v) <= 2002,
    'iyr': lambda v: 2010 <= int(v) <= 2020,
    'eyr': lambda v: 2020 <= int(v) <= 2030,
    'hgt': lambda v: (v.endswith('cm') and 150 <= int(v[:-2]) <= 193) or (v.endswith('in') and 59 <= int(v[:-2]) <= 76),
    'hcl': lambda v: COLOR_RE.fullmatch(v),
    'ecl': lambda v: v in ('amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'),
    'pid': lambda v: len(v) == 9 and v.isdigit(),
}

def one():
    passwords = read_passwords()
    passwords = filter(lambda password: all((field in password for field in REQUIRED_FIELDS)), passwords)
    print(len(list(passwords)))

def two():
    passwords = read_passwords()
    passwords = filter(lambda password: all((field in password for field in REQUIRED_FIELDS)), passwords)
    passwords = filter(lambda password: all((func(password[field]) for field, func in REQUIRED_FIELDS.items())), passwords)
    print(len(list(passwords)))

one()
two()
